<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import warnings



In [4]:
warnings.filterwarnings('ignore')

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [5]:
# текст запроса
query_3_1 = f'''
SELECT 
    COUNT(*)  "Count_vacancies" 
FROM
    public.vacancies
'''

In [6]:
# результат запроса
df_vacanies = pd.read_sql_query(query_3_1, connection)
df_vacanies

Count_vacancies
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
# текст запроса
query_3_2 = f'''
SELECT 
    COUNT(*) as "Count_employers"
FROM
    public.employers
'''

In [8]:
# результат запроса
df_employers = pd.read_sql_query(query_3_2, connection)
df_employers

Count_employers
0            23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query_3_3 = f'''
SELECT 
    COUNT(*) as "Count_regions"
FROM 
    public.areas
'''

In [10]:
# результат запроса
df_areas = pd.read_sql_query(query_3_3, connection)
df_areas

Count_regions
0           1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# текст запроса
query_3_4 = f'''
SELECT 
    COUNT(*) as "Count_fields_activity"
FROM 
    public.industries
'''

In [12]:
# результат запроса
df_industries = pd.read_sql_query(query_3_4, connection)
df_industries

Count_fields_activity
0                    294

***

Выводы по предварительному анализу данных:

* В базе содержится 49197  вакансий
* География представлена 1362 городами
* 23501 работателей, которые представлены  294 сферами деятельности 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
n = 5 # Количество  первых строк, которое вывведит на  экран
query_4_1 = f'''
SELECT 
    a.name as "Region",
    COUNT(v.id) as "Count_vacancies"
FROM 
    public.vacancies v
JOIN public.areas a on v.area_id = a.id      
GROUP BY a.id
ORDER BY count(v.id) DESC
LIMIT {n}     
'''

In [14]:
# результат запроса
df_cntarea = pd.read_sql_query(query_4_1, connection)
df_cntarea

Region  Count_vacancies
0           Москва             5333
1  Санкт-Петербург             2851
2            Минск             2112
3      Новосибирск             2006
4           Алматы             1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''
SELECT
    COUNT(v.id) as "Count_vacancies with salary"
FROM 
    public.vacancies v
WHERE  v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL      
'''

In [16]:
# результат запроса
df_salary = pd.read_sql_query(query_4_2, connection)
df_salary

Count_vacancies with salary
0                        24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = f'''
SELECT
    ROUND(AVG(v.salary_from)) as "Salary_from",
    ROUND(AVG(v.salary_to)) as "Salary_to"
FROM 
    public.vacancies v
      
'''

In [18]:
# результат запроса
salary_difference = pd.read_sql_query(query_4_3, connection)
salary_difference

Salary_from  Salary_to
0      71065.0   110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_4_4 = f'''
SELECT
    v.schedule as "Schedule",
    v.employment as "Employment",
    COUNT(v.id)  as " Count_vacancies schedule - employment"
FROM 
    public.vacancies v
GROUP BY v.schedule, v.employment
ORDER BY COUNT(v.id) DESC      
'''

In [20]:
# результат запроса
schedule_employment = pd.read_sql_query(query_4_4, connection)
schedule_employment

Schedule           Employment   
0        Полный день     Полная занятость  \
1   Удаленная работа     Полная занятость   
2      Гибкий график     Полная занятость   
3   Удаленная работа  Частичная занятость   
4     Сменный график     Полная занятость   
5        Полный день           Стажировка   
6     Вахтовый метод     Полная занятость   
7        Полный день  Частичная занятость   
8      Гибкий график  Частичная занятость   
9        Полный день     Проектная работа   
10  Удаленная работа     Проектная работа   
11     Гибкий график           Стажировка   
12    Сменный график  Частичная занятость   
13  Удаленная работа           Стажировка   
14     Гибкий график     Проектная работа   
15    Сменный график           Стажировка   
16    Вахтовый метод     Проектная работа   
17    Сменный график     Проектная работа   

     Count_vacancies schedule - employment  
0                                    35367  
1                                     7802  
2                                     1593  
3                                     1312  
4                                      940  
5                                      569  
6                                      367  
7                                      347  
8                                      312  
9                                      141  
10                                     133  
11                                     116  
12                                     101  
13                                      64  
14                                      18  
15                                      12  
16                                       2  
17                                       1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_5 = f'''
SELECT
    v.experience as "Experience",
    COUNT(v.id)  as " Count"
FROM 
    public.vacancies v
GROUP BY v.experience
ORDER BY COUNT(v.id)      
'''

In [22]:
# результат запроса
df_experience = pd.read_sql_query(query_4_5, connection)
df_experience

Experience   Count
0         Более 6 лет    1337
1           Нет опыта    7197
2       От 3 до 6 лет   14511
3  От 1 года до 3 лет   26152

***

Выводы по детальному анализу вакансий:
* Количество вакансий, где  указана заработная плата составляет 24073. Можно сделать вывод,  что оплата труда будет  определяться при собеседовании и оценки навыков соискателя работадателем. Основная часть работадателей ищут работников с минимальным опытом работы или без опыта предполагает,  что компании готовы вкладываться в развитие своих новых подчененых возможно за счет минимизации их зарплаты.
* Пятерка городов по   количеству  вакансий представлена крупными городами России, Белорусии и Казахстана.
* В основном ещут работников на полную занятость. Преобладают вакансии  Полный день - 	Полная занятость.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
# текст запроса
query_5_1 = f'''
(SELECT
    e.name as "Employer_name",
    COUNT(v.id) as "Count_employer"
FROM 
    public.vacancies v
JOIN public.employers e ON e.id = v.employer_id
GROUP BY e.id
ORDER BY COUNT(v.id) DESC
LIMIT 1)

UNION ALL

(SELECT
    e.name as "Employer_name",
    COUNT(v.id) as "Count_employer"
FROM 
    public.vacancies v
JOIN public.employers e ON e.id = v.employer_id
GROUP BY e.id
ORDER BY COUNT(v.id) DESC
OFFSET 4 LIMIT 1)
'''

In [24]:
# результат запроса
df_employers = pd.read_sql_query(query_5_1, connection)
df_employers

Employer_name  Count_employer
0         Яндекс            1933
1  Газпром нефть             331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
# текст запроса
query_5_2 = f'''
SELECT 
    a.name as "Name_area",
    COUNT(e.name) as "Count_employers",
    COUNT(v.id) as "Count_vacancies"
FROM 
    public.employers e
LEFT JOIN public.vacancies v ON e.area = v.area_id
JOIN public.areas a ON a.id  = e.area
GROUP BY a.name
ORDER BY COUNT(v.id) ASC, COUNT(e.name) DESC

'''

In [26]:
# результат запроса
df_areas_cntvacancies = pd.read_sql_query(query_5_2, connection)
df_areas_cntvacancies

Name_area  Count_employers  Count_vacancies
0                Россия              410                0
1             Казахстан              207                0
2    Московская область               75                0
3    Краснодарский край               19                0
4    Ростовская область               18                0
..                  ...              ...              ...
649         Новосибирск          1149438          1149438
650              Алматы          1364132          1364132
651               Минск          2354880          2354880
652     Санкт-Петербург          6320667          6320667
653              Москва         31272712         31272712

[654 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
# текст запроса
query_5_3 = f'''
SELECT 
    DISTINCT e.name as "Name_employer",
    COUNT(DISTINCT v.area_id) as "Count_regions"
FROM 
    public.employers e
JOIN public.vacancies v ON e.id = v.employer_id
GROUP BY e.name
ORDER BY  COUNT(DISTINCT v.area_id) DESC   
'''

In [28]:
# результат запроса
df_area_employers = pd.read_sql_query(query_5_3, connection)
df_area_employers

Name_employer  Count_regions
0                       Яндекс            181
1                   Ростелеком            152
2                   Спецремонт            116
3       Поляков Денис Иванович             88
4                    ООО ЕФИН              71
...                        ...            ...
14761     Яроцкий Е. В./ MoonX              1
14762  Ясли-сад №374 г. Минска              1
14763                      ЯСП              1
14764        Ястро-Переработка              1
14765                     ЯТСС              1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [29]:
# текст запроса

query_5_4 = f'''
SELECT
    COUNT(*) as "Count_employer"
FROM 
    public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL 
'''

In [30]:
# результат запроса
df_employers_industries = pd.read_sql_query(query_5_4, connection)
df_employers_industries

Count_employer
0            8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [31]:
# текст запроса
query_5_5 = f'''
SELECT
    e.name as "Name_employer",
    COUNT(ei.industry_id) as "Count_industries"
FROM 
    public.employers e
JOIN public.employers_industries ei ON e.id = ei.employer_id
GROUP BY e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY e.name
OFFSET 2 LIMIT 1 
'''

In [32]:
# результат запроса
df_cnt_industries = pd.read_sql_query(query_5_5, connection)
df_cnt_industries

Name_employer  Count_industries
0          2ГИС                 4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [33]:
# текст запроса
query_5_6 = f'''
SELECT
    COUNT(*) as "Count_employers"
FROM 
    public.industries i
JOIN public.employers_industries ei ON i.id = ei.industry_id
WHERE i.name = 'Разработка программного обеспечения' 
'''


In [34]:
# результат запроса
df_employer_software_development = pd.read_sql_query(query_5_6, connection)
df_employer_software_development

Count_employers
0             3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [35]:
# код для получения списка городов-милионников
# Получаем dataset городов-милионников
df_cities = pd.read_html('https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8')[1]
# Извлекаем из  базы список городов
cities = tuple(df_cities['Город'])


In [36]:
# текст запроса
query_5_7 = f'''
SELECT
    a.name as name_city,
    COUNT(v.id) as count_vacancies
FROM 
    public.vacancies v
JOIN public.employers e ON e.id = v.employer_id
JOIN public.areas a   ON a.id = v.area_id
WHERE a.name IN {cities} AND e.name = 'Яндекс'
GROUP BY a.name 

UNION ALL

SELECT
    'Total' ,COUNT(v.id)
FROM 
    public.vacancies v
JOIN public.employers e ON e.id = v.employer_id
JOIN public.areas a   ON a.id = v.area_id
WHERE a.name IN {cities} AND e.name = 'Яндекс'
ORDER BY 2
'''

In [37]:
# результат запроса
df_yandex = pd.read_sql_query(query_5_7, connection)
df_yandex

name_city  count_vacancies
0              Омск               21
1         Челябинск               22
2        Красноярск               23
3         Волгоград               24
4             Пермь               25
5            Казань               25
6    Ростов-на-Дону               25
7               Уфа               26
8            Самара               26
9         Краснодар               30
10          Воронеж               32
11      Новосибирск               35
12  Нижний Новгород               36
13     Екатеринбург               39
14  Санкт-Петербург               42
15           Москва               54
16            Total              485

***

Выводы по анализу работодателей:

* Большое количество вакансий представлены компаниями IT и банковских  сферах. Также крупные компании ищут работниках во многих регионах, что говорит о развитой филиальной  сети.
* Многие компании имеют несколько сфер  деятельности , что говорит  о развитии в нескольких направлениях и  поиска работников разных специальностей.
* 3553 компаний занимаются разработкой программного обеспечения
* У одного из лидеров IT сфере компания "Яндекс" представлены вакансии во всех крупных городах России. Можно предположить о постаянном росте компании или нехватки специалистов.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [38]:
# текст запроса
query_6_1 = f'''
SELECT
    COUNT(v.name) as "Count_vacancies"
FROM 
    public.vacancies v
WHERE v.name ILIKE '%data%' OR v.name ILIKE '%данн%'     
'''

In [39]:
# результат запроса
df_cnt_data = pd.read_sql_query(query_6_1, connection)
df_cnt_data

Count_vacancies
0             1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [40]:
# Оставим только те вакансии, которые относятся  к DS. В следующих заданиях будет только  с вакансиями Data Science
query_6 = f'''
SELECT 
    * 
FROM 
    public.vacancies v
WHERE (v.name ILIKE '%data scientist%'
    OR v.name ILIKE '%data science%'
    OR v.name ILIKE '%исследователь данных%'
    OR (v.name LIKE '%ML%' AND v.name NOT LIKE'%HTML%')
    OR v.name ILIKE '%machine learning%'
    OR v.name ILIKE '%машинн%обучен%')
    
'''

In [41]:
# текст запроса
query_6_2 = f'''
SELECT 
COUNT(*) as count_junior_vacancies
FROM 
    ({query_6}) as data_science
WHERE 
    name ILIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
'''

In [42]:
# результат запроса
df_cnt_junior = pd.read_sql_query(query_6_2, connection)
df_cnt_junior

count_junior_vacancies
0                      51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# текст запроса
query_6_3 = f'''
SELECT 
COUNT(*) as count_sql_vacancies
FROM 
    ({query_6}) as data_science
WHERE
    key_skills ILIKE '%SQL%'
    OR key_skills ILIKE '%postgres%'
'''    

In [44]:
# результат запроса
df_cnt_sql = pd.read_sql_query(query_6_3, connection)
df_cnt_sql

count_sql_vacancies
0                  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# текст запроса
query_6_4 = f'''
SELECT 
COUNT(*) as count_python_vacancies
FROM 
    ({query_6}) as data_science
WHERE 
    key_skills ILIKE '%Python%'
    
'''    

In [46]:
# результат запроса
df_cnt_python = pd.read_sql_query(query_6_4, connection)
df_cnt_python

count_python_vacancies
0                     351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
# текст запроса
query_6_5 = f'''
SELECT 
    ROUND(AVG(LENGTH (key_skills) - LENGTH (REPLACE(key_skills,  CHR(9), '')) + 1), 2) as "Avg_skills"
FROM 
    ({query_6}) as data_science
WHERE 
    key_skills IS NOT NULL
    
'''


In [48]:
# результат запроса
df_cnt_skills = pd.read_sql_query(query_6_5, connection)
df_cnt_skills

Avg_skills
0        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [49]:
# текст запроса
query_6_6 = f'''
SELECT 
    experience,
    ROUND(AVG(COALESCE((salary_from + salary_to)/2, salary_to, salary_from))) as "Avg_salary"  
FROM 
    ({query_6}) as data_science
WHERE 
    salary_to IS NOT NULL OR salary_from IS NOT NULL
GROUP BY experience   
'''

In [50]:
# результат запроса
ds_experience_salary = pd.read_sql_query(query_6_6, connection)
ds_experience_salary

experience  Avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

Выводы по предметному анализу:
* В данной базе представлены 1771 вакансий по работе с данными из них 480 по напралению Data Science
* Для молодых дата-сайентистов есть 51 предложение, где можно начать свою карьеру.
* Знания основных инструментов дата - сайентиста в вакансиях SQL - 201 предложение, Python - 301. Можно предположить Python более универсальный инструмент для работы данными.
* В предложениях о работе требования к специалистам указывают в среднем более 6 ключевых навыков, что говорит о том ,что знания только Python и SQL недостаточно и нужно развиваться в различных инструментариях.
* Заработная плата варьируется в зависимости от опыта от  75000 до 250000

# Общий вывод по проекту

Подведем итог исследования, обобщите выводы:
* В базе представлены 49197 вакансий от 23501 работадателей. Большинство предложений из крупных горов России и ближнего зарубежья.
* Большинство предложений это работа в офисе на постоянной основе и полный день. Большое количество вакансий представлены компаниями IT и банковских  сферах
* В основном ищут специалистов системного администрирования	и тех поддержки. 
* Больше половины вакансий ищут рабоников с маленьким опытом работы или без  опыта, а также у многих не укахзана заработная плата, можно сделать вывод что компании готовы вкладывать в развитие своих новых сотрудников, за счет первоночальной минимальной оплаты труда.
* Для начинающих специалистов по Data Science есть возможность выбора из 51 вакансии, но требования  к молодому специалисту Data Science к ключевым навыкм состовляет 7. Эти повышенные требования компенсируется  заработной платой.  Молодому специалисту без опыта работы Data Science предлогают в среднем 75000, хотя в среднем по  всей базе средняя у молодого специалиста 40000.

In [51]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

7.1 Определим какие вакансии требуются чаще

In [52]:
# текст запроса
query_7_1 = f'''
SELECT 
    v.name as "Name_vacancies",
    COUNT(v.id) as "Count_vacancies"
FROM 
    public.vacancies v
GROUP BY v.name
ORDER BY 2 DESC

'''

In [53]:
# результат запроса
df_count_vacancies = pd.read_sql_query(query_7_1, connection)
df_count_vacancies

Name_vacancies  Count_vacancies
0                     Системный администратор             1761
1                              Программист 1С             1153
2                             Бизнес-аналитик              488
3            Специалист технической поддержки              457
4                          Системный аналитик              432
...                                       ...              ...
21218            Генеральный менеджер Проекта                1
21219    Ведущий инженер-программист (1С:УПП)                1
21220       Главный специалист IT направления                1
21221  Сотрудник Отдела информационных систем                1
21222             Старший финансовый менеджер                1

[21223 rows x 2 columns]

7.2 Сколько ключевых навыков в среднем указывают в вакансиях  DS для специалистов уровня Junior?

In [54]:
# текст запроса
query_7_2 = f'''
SELECT 
    ROUND(AVG(LENGTH (key_skills) - LENGTH (REPLACE(key_skills,  CHR(9), '')) + 1), 2) as "Avg_skills"
FROM 
    ({query_6}) as data_science
WHERE 
    key_skills IS NOT NULL
    AND (name ILIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка')

'''

In [55]:
# результат запроса
df_skills_junior = pd.read_sql_query(query_7_2, connection)
df_skills_junior

Avg_skills
0        7.27

7.3 Запрос, позволяющий вычислить, какую зарплату  указывают для каждого типа требуемого опыта

In [56]:
# текст запроса
query_7_3 = f'''
SELECT 
    v.experience,
    ROUND(AVG(COALESCE((v.salary_from + v.salary_to)/2, salary_to, salary_from))) as "Avg_salary"  
FROM 
    public.vacancies v
WHERE 
    v.salary_to IS NOT NULL OR v.salary_from IS NOT NULL
GROUP BY v.experience
'''   

In [57]:
ds_salary = pd.read_sql_query(query_7_3, connection)
ds_salary

experience  Avg_salary
0         Более 6 лет    166256.0
1       От 3 до 6 лет    133211.0
2           Нет опыта     40472.0
3  От 1 года до 3 лет     76541.0

In [58]:
#connection.close()